src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -980,24 +977,47 @@ private void FreePinHandles()
                 if (spe != null)
                 {
-                    if (spe._buffer != null && spe._count > 0)
+                    if (spe.Buffer != null && spe.Count > 0)
                     {
                         // This element is a buffer.
-                        sendPacketsDescriptor[descriptorIndex].buffer = Marshal.UnsafeAddrOfPinnedArrayElement(spe._buffer, spe._offset);
-                        sendPacketsDescriptor[descriptorIndex].length = (uint)spe._count;
-                        sendPacketsDescriptor[descriptorIndex].flags = (Interop.Winsock.TransmitPacketsElementFlags)spe._flags;
+                        sendPacketsDescriptor[descriptorIndex].buffer = Marshal.UnsafeAddrOfPinnedArrayElement(spe.Buffer, spe.Offset);
+                        sendPacketsDescriptor[descriptorIndex].length = (uint)spe.Count;
+                        sendPacketsDescriptor[descriptorIndex].flags =
+                            Interop.Winsock.TransmitPacketsElementFlags.Memory | (spe.EndOfPacket
+                                ? Interop.Winsock.TransmitPacketsElementFlags.EndOfPacket
+                                : 0);
                         descriptorIndex++;
                     }
-                    else if (spe._filePath != null)
+                    else if (spe.FilePath != null)
                     {
                         // This element is a file.
                         sendPacketsDescriptor[descriptorIndex].fileHandle = _sendPacketsFileStreams[fileIndex].SafeFileHandle.DangerousGetHandle();
-                        sendPacketsDescriptor[descriptorIndex].fileOffset = spe._offset;
-                        sendPacketsDescriptor[descriptorIndex].length = (uint)spe._count;
-                        sendPacketsDescriptor[descriptorIndex].flags = (Interop.Winsock.TransmitPacketsElementFlags)spe._flags;
+                        sendPacketsDescriptor[descriptorIndex].fileOffset = spe.LongOffset;
+                        sendPacketsDescriptor[descriptorIndex].length = (uint)spe.Count;
+                        sendPacketsDescriptor[descriptorIndex].flags =
+                            Interop.Winsock.TransmitPacketsElementFlags.File | (spe.EndOfPacket
+                                ? Interop.Winsock.TransmitPacketsElementFlags.EndOfPacket
+                                : 0);
                         fileIndex++;
                         descriptorIndex++;
                     }
+                    else if (spe.FileStream != null)
+                    {
+                        // This element is a file stream.
+                        sendPacketsDescriptor[descriptorIndex].fileHandle = spe.FileStream.SafeFileHandle.DangerousGetHandle();
+                        sendPacketsDescriptor[descriptorIndex].fileOffset = spe.LongOffset;
+
+                        // Workaround from reading stream from current position
+                        if (spe.LongOffset == 0)
+                            spe.FileStream.Seek(0, SeekOrigin.Begin);
```

Comment:
I don't understand this. Why are we changing the FileStream?

Reply:
I added seek tests and them failed if offset equal to zero for `FileStream` packets.
I do not know why but if offset equal to zero windows send packet from current `FileStream` position.
At documentation https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms740570(v=vs.85).aspx wrote that offset should be -1 to read from current position but it seem windows read from current position when offset equal 0 also.

This is exactly the case with the first package. I will make tests for non fists packages.

Reply:
> I do not know why but if offset equal to zero windows send packet from current FileStream position.
At documentation https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms740570(v=vs.85).aspx wrote that offset should be -1 to read from current position but it seem windows read from current position when offset equal 0 also.

Hmm.. that is very unexpected. We can't change the FileStream to workaround that, as that isn't thread safe.
Do you know if the TransmitPackets API may be used concurrently with the same file handle?

@geoffkizer are you familiar with this API?

Reply:
> Do you know if the TransmitPackets API may be used concurrently with the same file handle?

FileStream support is new for TransmitPackets API.

Reply:
My question is about the Windows OS TransmitPackets function (what we are using here to build the .NET API).

Reply:
> Do you know if the TransmitPackets API may be used concurrently with the same file handle?

I think no. The same file handle can not be used concurrently without synchronization.

Reply:
> I think no. The same file handle can not be used concurrently without synchronization.

If that is true, that seriously limits the usability of the FileStream overload. From the API proposal, only the ability to set specific options remains, which is a very specific use-case.

We need to better understand TransmitPackets to know how to proceed.

Reply:
The way things usually work in Win32 is that file handles opened for async IO (via `FILE_FLAG_OVERLAPPED`) do not maintain the current file position and thus are safe to be used concurrently. That said, I have no idea if `FileStream` opens files using `FILE_FLAG_OVERLAPPED`. Unfortunately `FileStream` is a somewhat poor abstraction for Win32's async IO facilities...

Reply:
FileStream will use FILE_FLAG_OVERLAPPED if it's opened with a useAsync:true or FileOptions.Asynchronous ctor argument.

Reply:
Regardless, I don't think we should be doing `spe.FileStream.Seek(0, SeekOrigin.Begin);`.  If there's an odd behavior being observed, we should figure out why and address the root cause rather than working around it like this.

Reply:
@stephentoub Windows change stream position when sending packets so I think that not issue that we change position before call to sending packets because stream position will be changed any way.

Reply:
This behavior should have been clarified during design. It's either that the stream position is updated or not but as already mentioned Windows may update or not the file pointer depending on how the file handle was created. Which incidentally will make this feature problematic to implement...

Reply:
@mikedn Ok I checks this and will create test to protect Position changing.